blackbird_python/blackbird/program.py

```diff
@@ -0,0 +1,290 @@
+# Copyright 2019 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# pylint: disable=too-many-return-statements,too-many-branches,too-many-instance-attributes
+"""
+Blackbird Program class
+=======================
+
+**Module name:** `blackbird.program`
+
+.. currentmodule:: blackbird.program
+
+This module contains a Python class representing a Blackbird
+program using standard Python data types.
+
+The functions :func:`~.load`, and :func:`~.loads` will read Blackbird scripts
+and return an instance of the :class:`BlackbirdProgram` class.
+
+Summary
+-------
+
+.. autosummary::
+    numpy_to_blackbird
+    BlackbirdProgram
+
+Code details
+~~~~~~~~~~~~
+"""
+import numpy as np
+
+
+def numpy_to_blackbird(A, var_name):
+    """Converts a numpy array to a Blackbird script array type.
+
+    Args:
+        A (array): 2-dimensional NumPy array
+        var_name (str): the array variable name
+
+    Returns:
+        list[str]: list containing each line representing the
+            Blackbird array variable declaration
+    """
+
+    if np.any(np.iscomplex(A)):
+        # complex array
+        script = ["complex array {}[{}, {}] =".format(var_name, *A.shape)]
+        for row in A:
+            row_str = "    " + ", ".join(
+                ["{0}{1}{2}j".format(n.real, "+-"[int(n.imag < 0)], abs(n.imag)) for n in row]
+            )
+            script.append(row_str)
+
+    elif np.all(np.mod(A, 1) == 0):
```

Comment:
This will register an array which only contains floats 0.0 and 1.0 as an int array, correct? Any use case where the user would want these to remain floats?

Reply:
Actually this should work on any integer array (so 2.0 -> 2 etc).

Maybe I'm trying to be too clever here. Should I just use the numpy.dtype as the 'truth' for the array type?

Reply:
Probably. At best, we're saving a few bits